packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -1157,9 +1157,20 @@ class SliverMultiBoxAdaptorElement extends RenderObjectElement implements Render
         final Key key = _childElements[index].widget.key;
         final int newIndex = key == null ? null : widget.delegate.findIndexByKey(key);
         if (newIndex != null && newIndex != index) {
+          // The layout offset of the child being moved is no longer accurate.
+          final SliverMultiBoxAdaptorParentData childParentData = _childElements[index].renderObject?.parentData;
+          if (childParentData != null)
+            childParentData.layoutOffset = null;
+          // Inherits layoutOffset if there is an existing element in the
+          // destination
+          if (_childElements[newIndex] != null) {
+            final SliverMultiBoxAdaptorParentData newParentData = _childElements[newIndex]?.renderObject?.parentData;
+            if (childParentData != newParentData && childParentData != null && newParentData != null) {
+              childParentData.layoutOffset = newParentData.layoutOffset;
+            }
+          }
           newChildren[newIndex] = _childElements[index];
-          // We need to make sure the original index gets processed.
-          newChildren.putIfAbsent(index, () => null);
```

Comment:
Another important change is it no longer forces the the child in the old index to be repopulate. It will create a child with null layout offset. The old index will be repopulate in performlayout if it is needed.

Reply:
There will be no unintended scrolling.
If this index suppose to be first child and is now missing, perform layout will try to recreate and relayout this child from other existing child, and eventually it will figure out  there is no need to correct scroll offset.

But now i think of it, I might be able to refactor a little bit to make it work. It will just make the algorithm slightly more complicated.